docs(frontend): adding a use-case for private information retrieval

bcm-at-zama opened the pull request:
Work in progress, I've made experiments for PIR with Concrete. I think we can find use-cases for which it's not that bad.

umut-sahin commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,509 @@
+   "outputs": [],
+   "source": [
+    "@fhe.compiler({\"one_hot_vector\": \"encrypted\", \"database\": \"clear\"})\n",
+    "def query(one_hot_vector: np.array, database: np.array) -> int:\n",
+    "    return np.dot(one_hot_vector, database)"
```
I think `index_into_database` or `get_ith_element_of_database` would be a better name to describe the operation. `query` sounds like you want to compute `i in database` instead of `database[i]`.

bcm-at-zama replied:
Renamed to get_ith_element_of_database

umut-sahin reviewed (changes requested):
Could you create a `.py` script as well. It'll be necessary for testing and benchmarks. Maybe even do those directly, wdyt?

Very cool example btw!

bcm-at-zama commented:
Let me have a look to make a .py yes. I am currently installing the new computer and I have an issue running the nb: it crashes when doing the 

encrypted_y = circuit.run(encrypted_x, database)

I am investigating.

umut-sahin commented:
Could you share it in `#concrete` channel? I don't remember the issue, maybe someone can help.

bcm-at-zama commented:
I think I fixed that with https://github.com/zama-ai/concrete-ml/pull/870

bcm-at-zama commented on frontends/concrete-python/examples/pir/README.md:
```diff
@@ -0,0 +1,3 @@
+# Private information retrieval
+
+In this directory, we present some experiments about PIR made with Concrete, as well as a use-case to access a database to ask if a phone number is a spam, without revealing the phone number. It can be compared with Apple's recent [study and open-source code](https://github.com/apple/live-caller-id-lookup-example).
```
ok done

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,509 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "x = make_one_hot_vector(0, 5)\n",
+    "assert np.array_equal(x, np.array([1, 0, 0, 0, 0]))\n",
+    "\n",
+    "x = make_one_hot_vector(2, 7)\n",
+    "assert np.array_equal(x, np.array([0, 0, 1, 0, 0, 0, 0]))"
```
ok done

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,509 @@
+      "--------------------------------------------------------------------------------\n",
+      "module {\n",
+      "  func.func @main(%arg0: tensor<8x!FHE.eint<5>>, %arg1: tensor<8xi6>) -> !FHE.eint<5> {\n",
+      "    %0 = \"FHELinalg.to_signed\"(%arg0) : (tensor<8x!FHE.eint<5>>) -> tensor<8x!FHE.esint<5>>\n",
+      "    %1 = \"FHELinalg.dot_eint_int\"(%0, %arg1) : (tensor<8x!FHE.esint<5>>, tensor<8xi6>) -> !FHE.esint<5>\n",
+      "    %2 = \"FHE.to_unsigned\"(%1) : (!FHE.esint<5>) -> !FHE.eint<5>\n",
+      "    return %2 : !FHE.eint<5>\n",
```
```
Computation Graph
--------------------------------------------------------------------------------
%0 = one_hot_vector        # EncryptedTensor<uint1, shape=(8,)>        ∈ [0, 1]
%1 = database              # ClearTensor<uint5, shape=(8,)>            ∈ [2, 21]
%2 = dot(%0, %1)           # EncryptedScalar<uint5>                    ∈ [2, 21]
return %2
--------------------------------------------------------------------------------

MLIR
--------------------------------------------------------------------------------
module {
  func.func @main(%arg0: tensor<8x!FHE.eint<5>>, %arg1: tensor<8xi6>) -> !FHE.eint<5> {
    %0 = "FHELinalg.to_signed"(%arg0) : (tensor<8x!FHE.eint<5>>) -> tensor<8x!FHE.esint<5>>
    %1 = "FHELinalg.dot_eint_int"(%0, %arg1) : (tensor<8x!FHE.esint<5>>, tensor<8xi6>) -> !FHE.esint<5>
    %2 = "FHE.to_unsigned"(%1) : (!FHE.esint<5>) -> !FHE.eint<5>
    return %2 : !FHE.eint<5>
  }
}
--------------------------------------------------------------------------------
```

@umut-sahin

umut-sahin replied:
Okay, this shouldn't happen, I'll take a look.

umut-sahin replied:
https://github.com/zama-ai/concrete/pull/1047

bcm-at-zama replied:
saw that thx

bcm-at-zama commented on frontends/concrete-python/tests/execution/test_examples.py:
```diff
@@ -271,3 +271,37 @@ def test_levenshtein_distance_randomly(alphabet_name, max_length, helpers):
         actual_distance = levenshtein_distance.calculate(str1, str2, "simulate", show_distance=True)
         assert actual_distance == expected_distance
+
+def test_pir_basics():
```
@umut-sahin I'm trying to test without being able to run that locally! If you can help me having something working locally, it's awesome. Or a way to test simply & quickly on the CI.

bcm-at-zama commented on frontends/concrete-python/Makefile:
```diff
@@ -69,6 +69,13 @@ tfhers-utils:
 
 pytest: pytest-default
 
+pytest-benoit: tfhers-utils
```
trying something to test in the CI quickly

bcm-at-zama commented:
@umut-sahin is it better now? I have adressed your comments I think + added some tests in pytest. I haven't done the benchmark part, I am less sure how to do it but maybe we can do it together?

bcm-at-zama commented on .github/workflows/concrete_python_tests_linux.yml:
```diff
@@ -116,7 +116,7 @@ jobs:
             
             mkdir ./KeySetCache
-            KEY_CACHE_DIRECTORY=./KeySetCache make pytest
+            KEY_CACHE_DIRECTORY=./KeySetCache make pytest-benoit
```
when the PR is good for you @umut-sahin I remove this

bcm-at-zama commented on frontends/concrete-python/Makefile:
```diff
@@ -67,6 +67,13 @@ tfhers-utils:
 
 pytest: pytest-default
 
+pytest-benoit: tfhers-utils
```
when the PR is good for you @umut-sahin I remove this

umut-sahin commented:
Tests seem to fail, could you fix it before I give another review?

bcm-at-zama commented:
> Tests seem to fail, could you fix it before I give another review?

sure my bad, sorry for that, I think they worked locally but let me see

bcm-at-zama commented:
@umut-sahin : 
- I have fixed the test (but it would be nicer with a way to reset the compilation, ie https://github.com/zama-ai/concrete-internal/issues/845)
- I have added a small thing we have discussed yersterday with @aquint-zama and @BourgerieQuentin : having an hash-based system as well to have eg URL filtering. That's in the end of the nb

bcm-at-zama commented:
CC @aquint-zama and @BourgerieQuentin if you want to have a look and review a bit the notebook

umut-sahin commented:
You'll have the ability to reset pretty soon :wink: 
https://github.com/zama-ai/concrete/pull/1048

bcm-at-zama commented:
@umut-sahin the tests look good!

BourgerieQuentin commented on frontends/concrete-python/examples/pir/pir_utils.py:
```diff
@@ -0,0 +1,55 @@
+
+
+def make_one_hot_vector(index: int, size: int) -> np.ndarray:
+
```
ws?

bcm-at-zama replied:
I don't know what it mean

bcm-at-zama replied:
removing this file

BourgerieQuentin commented on frontends/concrete-python/examples/pir/pir_utils.py:
```diff
@@ -0,0 +1,55 @@
+
+def make_one_hot_vector(index: int, size: int) -> np.ndarray:
+
+    answer = np.zeros(shape=(size,), dtype=np.int8)
```
is that dtype is madatory?

BourgerieQuentin replied:
It's more query than answer, right?

bcm-at-zama replied:
- if we don't have dtype, i think it takes float, no?
- "It's more query than answer, right?": I have not understood

bcm-at-zama replied:
removing this file

BourgerieQuentin commented on frontends/concrete-python/examples/pir/pir_utils.py:
```diff
@@ -0,0 +1,55 @@
+
+def compile_function(database, show_mlir=False, show_graph=False):
+
+    get_ith_element_of_database.reset()
```
Why the compiler need to be toplevel?

bcm-at-zama replied:
to be accessible in the pytest files

BourgerieQuentin commented on frontends/concrete-python/examples/pir/pir_utils.py:
```diff
@@ -0,0 +1,55 @@
+    get_ith_element_of_database.reset()
+
+    database_length = database.shape[0]
+    inputset_length = 100
```
Could we simplify that by computing the worst case query?

bcm-at-zama replied:
what does that mean

umut-sahin reviewed (changes requested):
I think there shouldn't be a `pir_utils.py`, there should be `pir.py`, which should be runable on it's own (e.g., to print an example, or assert some things, show timings).

Within it, there should be a class named `PrivateInfomationRetrival`, it should accept `configuration: Optional[fhe.Configuration] = None` and `compiled: bool = True` for its constructor. 

Lastly, the test logic should be defined in `tests` not in the example itself.

Sorry it's not a small request, but it's for being consistent across the examples!

umut-sahin commented:
I'm just saying we don't have `example_utils.py` anywhere else, and we try to provide a simple python script that the users can run directly. I think the structure of an example should be:
- start by having a plain python script to implement the algorithm, show how to use it, and do a few assertions to show how it works (so not for testing)
- then add tests for it using the class defined in the example script
- then add benchmarks for it using the class defined in the example script
- then optionally create a notebook to convert it from an example to a tutorial
The ordering is important here. We shouldn't neglect the others and directly be a tutorial instead of an example.

bcm-at-zama commented:
> I'm just saying we don't have `example_utils.py` anywhere else, and we try to provide a simple python script that the users can run directly. I think the structure of an example should be:
> 
> * start by having a plain python script to implement the algorithm, show how to use it, and do a few assertions to show how it works (so not for testing)
> * then add tests for it using the class defined in the example script
> * then add benchmarks for it using the class defined in the example script
> * then optionally create a notebook to convert it from an example to a tutorial
>   The ordering is important here. We shouldn't neglect the others and directly be a tutorial instead of an example.

I like it for the first three parts, but then I really think it's at the price of having nice notebook-tutorials. Those notebook-tutorials are really easier for the users, especially the beginners, to follow and appreciate.

bcm-at-zama commented:
We're not in the same view, let's see with another person to break the tie. @BourgerieQuentin , what do you think please? I am fine refactoring all of this, but I don't think it goes for the best of the final users.

bcm-at-zama commented:
I am starting a `make jupyter`, see https://github.com/zama-ai/concrete/pull/1056, we really need to test the notebooks (2 of the current 4's are broken!)

bcm-at-zama commented:
@BourgerieQuentin , we debated a bit with @umut-sahin and came to the following proposal:

To test the CP examples, we would either do it:
  - in pytest for the .py
  - or via juputer.sh for the .nb (it's a PR in progress, https://github.com/zama-ai/concrete/pull/1056)

What do you think of this proposal? 

An alternative approach was:

> We write the tutorials in a notebook, and convert it to a plain script, which we can just run, import from tests/benchmarks. CI could check if the conversion is correct. Wdyt?
During development, we can start with the notebook. When it's ready we can do:
TARGET=game_of_life make tutorial-to-example
Which would create:
examples/game_of_life/game_of_life.py
And in the CI, we'd check if this committed file is exactly the same as the notebook

but we thought it would be a bit over engineering to do that.

BourgerieQuentin commented:
Ok fine for me :+1:

bcm-at-zama commented:
Awesome so I'm going to remove my changes in *.py, and just keep the notebook, which will be tested with the other PR.

bcm-at-zama commented:
So now, it's just the notebook to review please.

BourgerieQuentin commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+    "        (make_one_hot_vector(np.random.randint(database_length), database_length), database)\n",
+    "        for _ in range(inputset_length)\n",
+    "    ]\n",
+    "    circuit = get_ith_element_of_database.compile(inputset, show_mlir=show_mlir, show_graph=show_graph)\n",
```
`**1` instantiate the fhe.Compiler here instead of using decorator.

`circuit = fhe.Compiler(get_ith_element_of_database, {"one_hot_vector": "encrypted", "database": "clear"})`

bcm-at-zama replied:
lol, even you have forgotten about this syntax: from the README, it's two steps

```
compiler = fhe.Compiler(add, {"x": "encrypted", "y": "encrypted"})
circuit = compiler.compile(inputset)
```

I would have prefered to go for the new .reset but ok!

BourgerieQuentin commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+    "\n",
+    "for database_input_bits, database_output_bits in sample_list:\n",
+    "\n",
+    "    # FIXME: remove once I know how to reset\n",
```
So you can remove the def of the fhe.compiler here. `**1`

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+    "\n",
+    "Our database will be represented as a table T, and inputs to T will be `n_input_bits` bits. Cells of T are represented as `m_output_bits`.\n",
+    "\n",
+    "Before querying the database, the input `i` will be represented as one-hot vectors, ie a vector of `2**n_input_bits` bits: all of them are 0, but the one in the i-th position."
```
ok done

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "database_length = 8\n",
```
ok done

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+    "\n",
+    "assert database_output_bits == 5\n",
+    "\n",
+    "assert get_ith_element_of_database(make_one_hot_vector(0, size=database_length), database) == database[0]\n",
```
ok I have added a sentence

```
# For now, we have not compiled our functions so here, all the computations 
# in the following asserts are done in the clear, just to check the semantic
# of the functions
```

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+      "--------------------------------------------------------------------------------\n",
+      "module {\n",
+      "  func.func @main(%arg0: tensor<8x!FHE.eint<5>>, %arg1: tensor<8xi6>) -> !FHE.eint<5> {\n",
+      "    %0 = \"FHELinalg.to_signed\"(%arg0) : (tensor<8x!FHE.eint<5>>) -> tensor<8x!FHE.esint<5>>\n",
```
I'll let it to @umut-sahin

umut-sahin replied:
It can be anything.

bcm-at-zama replied:
Let me close this and merge, we can finish those conversation outside, and if needed, re-run the notebook with a new wheel

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+    }
+   ],
+   "source": [
+    "def compile_function(database, show_mlir=False, show_graph=False):\n",
```
good idea yes, thx, I should have done this before

bcm-at-zama replied:
could you check it's ok now?

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+   "source": [
+    "def compile_function(database, show_mlir=False, show_graph=False):\n",
+    "    database_length = database.shape[0]\n",
+    "    inputset_length = 100\n",
```
So:
- you're right it could be wrong
- you're right this single one could be sufficient
- but I don't like us advertising about single-element inputset, since people may try to do the same when it's not a good idea

so I added a:

```
# Also add the extreme value, which would actually be sufficient alone, as it reaches the maximal
    # values in get_ith_element_of_database
    inputset.append((make_one_hot_vector(np.argmax(database), database_length), database))
 ```

Good for you?

BourgerieQuentin replied:
Yes good for me :pray:

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+   ],
+   "source": [
+    "def compile_function(database, show_mlir=False, show_graph=False):\n",
+    "    database_length = database.shape[0]\n",
```
ok I added a `assert database.ndim == 1`
(yours doesn't work)

BourgerieQuentin replied:
great, I'm not a python master :laughing:

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@fhe.compiler({\"one_hot_vector\": \"encrypted\", \"database\": \"clear\"})\n",
```
ok done, see main one which I let open for you

bcm-at-zama commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+   "source": [
+    "def make_one_hot_vector(index: int, size: int) -> np.ndarray:\n",
+    "\n",
+    "    answer = np.zeros(shape=(size,), dtype=np.int8)\n",
```
I go for `one_hot_vector`. I find 'query' too much like an entry, not like an output

bcm-at-zama replied:
done

BourgerieQuentin commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,576 @@
+      "MLIR\n",
+      "--------------------------------------------------------------------------------\n",
+      "module {\n",
+      "  func.func @main(%arg0: tensor<8x!FHE.eint<5>>, %arg1: tensor<8xi6>) -> !FHE.eint<5> {\n",
```
still not understand why there are are signed conversion here @umut-sahin ? For me we should have just the dot product here..

umut-sahin replied:
It should be fixed after https://github.com/zama-ai/concrete/pull/1047. I think Benoit is running from pre-installed CP, which doesn't come with this change. Could you update to the latest nightly @bcm-at-zama?

bcm-at-zama replied:
```
commit a8bba20a7e9c3be24d9cd87fc6af660e6545e4e8
Author: Umut <umutsahin@protonmail.com>
Date:   Thu Sep 12 10:33:04 2024 +0300

    fix(frontend): remove unnecessary signed conversions
```

Maybe not!

bcm-at-zama replied:
I still have

MLIR
--------------------------------------------------------------------------------
module {
  func.func @main(%arg0: tensor<8x!FHE.eint<5>>, %arg1: tensor<8xi6>) -> !FHE.eint<5> {
    %0 = "FHELinalg.to_signed"(%arg0) : (tensor<8x!FHE.eint<5>>) -> tensor<8x!FHE.esint<5>>
    %1 = "FHELinalg.dot_eint_int"(%0, %arg1) : (tensor<8x!FHE.esint<5>>, tensor<8xi6>) -> !FHE.esint<5>
    %2 = "FHE.to_unsigned"(%1) : (!FHE.esint<5>) -> !FHE.eint<5>
    return %2 : !FHE.eint<5>
  }
}
--------------------------------------------------------------------------------

but maybe Concrete is too old?

BourgerieQuentin commented on frontends/concrete-python/examples/pir/PIR.ipynb:
```diff
@@ -0,0 +1,577 @@
+    "        [np.random.randint(2**database_output_bits) for _ in range(database_length)]\n",
+    "    )\n",
+    "\n",
+    "    circuit = compile_function(database, show_mlir=False, show_graph=False)\n",
```
so maybe just remove kwargs as it's default one

bcm-at-zama replied:
I would keep them just for the user to know they exist!

bcm-at-zama commented:
Thanks a lot @umut-sahin  / @BourgerieQuentin , let's merge!
